akka-cluster-sharding/src/main/scala/akka/cluster/sharding/SliceRangeShardAllocationStrategy.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2024-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding
+
+import scala.concurrent.Future
+
+import akka.actor.ActorRef
+import akka.actor.ActorRefScope
+import akka.actor.ActorSystem
+import akka.actor.MinimalActorRef
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent.CurrentClusterState
+import akka.cluster.Member
+import akka.cluster.sharding.ShardCoordinator.ActorSystemDependentAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin
+
+object SliceRangeShardAllocationStrategy {
+  // Not using Persistence because akka-persistence dependency could be optional
+  val NumberOfSlices = 1024
+
+  final class ShardBySliceMessageExtractor // FIXME
+}
+
+class SliceRangeShardAllocationStrategy(rebalanceLimit: Int)
+    extends ActorSystemDependentAllocationStrategy
+    with ClusterShardAllocationMixin {
+  import SliceRangeShardAllocationStrategy._
+
+  private var cluster: Cluster = _
+
+  override def start(system: ActorSystem): Unit = {
+    cluster = Cluster(system)
+  }
+
+  override protected def clusterState: CurrentClusterState = cluster.state
+  override protected def selfMember: Member = cluster.selfMember
+
+  override def allocateShard(
+      requester: ActorRef,
+      shardId: ShardId,
+      currentShardAllocations: Map[ActorRef, IndexedSeq[ShardId]]): Future[ActorRef] = {
+    val slice = shardId.toInt
+    if (slice >= NumberOfSlices)
+      throw new IllegalArgumentException("slice must be between 0 and 1023. Use `ShardBySliceMessageExtractor`.")
+
+    // sort regions by member age because if a node is added (random address) we don't want rebalance more than necessary
+    val regionsByMbr = regionsByMember(currentShardAllocations.keySet)
+    val regions = regionsByMbr.keysIterator.toIndexedSeq.sorted(Member.ageOrdering).map(regionsByMbr(_))
+    val rangeSize = NumberOfSlices / regions.size
+    val i = slice / rangeSize
```

Comment:
A problem with this first naive approach is that adding/removing nodes will reshuffle many shards. Consistent hashing would be nice, but since it's (dynamic) ranges I don't see how that can be used. I have an idea that it can instead find existing adjacent shards and prefer allocation to same region.

Reply:
Haven't tracked what the updated algorithm is yet, but feels like this issue could always be there in some form.

The optimal allocation (however it ends up there) will have all the shards for a slice range together. When these need to be reallocated, the optimal is to allocate as many as possible to one node again, but that either leads to reshuffles, or needing to accept unbalanced distributions, or they need to be redistributed over the other nodes creating fragmentation for the slice range.

But maybe in practice the find-neighbours approach naturally ends up balancing the tradeoff over time?